teps/0077-pipeline-partial-execution.md

```diff
@@ -0,0 +1,496 @@
+   and inability to retry would require a user to start the entire pipeline
+   over. A manual retry, with the ability to specify what tasks should
+   be skipped, may be helpful.
+2. Partial execution of pipeline. This is useful for reusing an existing
+   pipeline i.e. a user can disable certain task from an existing pipeline
+   and in this way run it without creating a new pipeline.
+3. Partial execution is also helpful for testing, i.e. skipping some tasks
+   and developing and testing iteratively and quickly.
+4. Pause and resume, i.e. one could manually cancel a running `PipelineRun`
+   and resume at later point.
+
+## Requirements
+
+- Create a new `PipelineRun` to resume or retry a completed `PipelineRun`.
+- Ability to partially run a Pipeline, without editing its definition or without
+  an existing `PipelineRun`.
+- An end user may not have to figure out the execution DAG, to be able to resume
+  or retry.
+
+## Proposal
+
+### Requesting API Changes
+1. Add `pipelineRunRef` under `PipelineRun.spec`. It has following fields: 
+   - `pipelineRunRef.name` which is the name of previously run `PipelineRun`.
+   - `pipelineRunRef.enableTasks` accepts an array of task names under it.
+2. Add `disableTasks` under `PipelineRun.spec`, which accepts an array
+    of task name.
+   - `name`: Name of the task to be disabled.
+
+Q. Why do we need `pipelineRunRef` when we have `disableTasks`? 
+
+`disableTasks` can be used to explicitly disable tasks that a user 
+do not wish to run. On the other hand, in `pipelineRunRef` tekton controller
+automatically figures out the tasks failed and unfinished, because it knows the
+DAG. For the end user, it can be difficult to figure out the DAG and prepare
+the accurate execution plan for the next pipeline run.
+
+Both, `pipelineRunRef` and `disableTasks` are optional fields. See examples
+1 and 2 below.
+
+### Semantics of execution.
+
+- `pipelineRunRef` : pipelineRunRef references a previous pipelineRun and by default
+    selects all the failed and unfinished tasks eligible for retrying/resuming.
+    It references results of completed tasks from previous run.
+
+- `pipelineRunRef.enableTasks`: If a task was successful in previous run, but
+    it is required by the current run, this section can be used to explicitly
+    enable it. For example, a task may perform some initialization for the
+    other tasks in `PipelineRun`.
+
+**Case: disabled task exists somewhere in the middle of DAG execution.**
+
+Consider the case of following failed `PipelineRun`.
+```
+F   A     D
+| \ |     |
+v   v     v 
+G   B --> E
+    |
+    v
+    C
+```
+
+1. 
+  - Previous Run stats from `prev-pipeline-run` :
+    - Successful Tasks: A 
+    - Failed Tasks: F, B
+    - Not yet started tasks: D, E, G, C
+  - Current Run: `new-pipeline-run`
+      - Disabled task: B, G
+      - To be executed: F, D, C
+
+Since task B is disabled, task E will also be disabled as it depends on B.
```

Comment:
what does "disabled" mean here? is it "not executed"? is that the same as "skipped"? I'm wondering if we could use more precise language here instead of "disable"

also, do we anticipate a scenario where `task B` is "disabled" but a user wants to execute the dependent `task E` (maybe it's an ordering dependency instead of resource dependency) -- this came up and we had to handle it in skipping strategies

Reply:
I chose, disable over skipped for the following reason:

When we support the feature discussed in `#Future work` i.e.  a task could be `disabled` but not its dependents (results could be pre-populated too). I could be wrong, but to me disabled seemed to be a bit precise ? Anyway, if you agree for future expansion of its functionality `skipped` is a more precise language. Happy to change it to skipped ! 👍

Reply:
that same reason applies to skipped since 0.27, a task could be `skipped` but not its dependents 

just wanted to flesh out what the difference is between `disabled` and `skipped` in terms of execution so we're not using different terms for the same thing 

also open to using `disabled` if that's clearer to people